### PR TITLE
Changes to support TLS access emitted in ILC (x64 win/linux)

### PR DESCRIPTION
**Pull request by VSadov**
These are companion changes for https://github.com/dotnet/runtime/pull/87148

The change is for win-x64 and linux-x64 combinations as that is enough to discuss and test the parts of the runtime design that are not platform-specific. 

More platform combinations can come as separate changes.

### CODE REVIEW

**Comment by VSadov** [comment]
CC: @filipnavara

**Review by filipnavara — APPROVED** [review]

**Review by jkotas — APPROVED** [review]

File: llvm/include/llvm/MC/MCObjectFileInfo.h [hunk 1]
```diff
@@ -344,7 +344,7 @@ class MCObjectFileInfo {
   }
 
   MCSection *getTLSExtraDataSection() const { return TLSExtraDataSection; }
-  const MCSection *getTLSDataSection() const { return TLSDataSection; }
+  MCSection *getTLSDataSection() const { return TLSDataSection; }
```
Comment (MichalStrehovsky) [inline llvm/include/llvm/MC/MCObjectFileInfo.h:347]: Why is this needed? Could we just mark things const on the objwriter side? Changes to LLVM outside of tools/objwriter need to be ported every time we update LLVM so there's a lot of value in keeping them as minimal as possible.

Comment (VSadov) [reply llvm/include/llvm/MC/MCObjectFileInfo.h:347]: > Could we just mark things const on the objwriter side

I tried and propagating the const requires a lot of changes (const can be viral). 
It is also inconsistent with the treatment other section accessors - i.e why getTLSDataSection is a const and getTLSBSSSection isn't? 

I think the difference is unintentional. If we want const, we should make other sections accessors const as well.

Comment (VSadov) [reply llvm/include/llvm/MC/MCObjectFileInfo.h:347]: One possibility is to keep `getTLSDataSection()` as `const` (for no reason) and just cast the const away in the object writer.  

There is a danger that something may eventually make dependency on this being const and it would not be easily noticeable when updating.

Comment (VSadov) [reply llvm/include/llvm/MC/MCObjectFileInfo.h:347]: > cast the const away in the object writer

Such workaround stands out in the code and feels like a mildly dangerous hack. 
I think we can do such change if removing the `const` becomes a burden when updating. It probably won't if upstream code does not change. It does not look like this part changes often.

Comment (VSadov) [reply llvm/include/llvm/MC/MCObjectFileInfo.h:347]: Just to be sure - I have noted the concern, but any way to deal with this issue has a downside. Removing the 'const' seems not distinctly worse than other solutions.  
We need to pick one way to move forward while decision is not binding, since it is an implementation detail that can be changed. I am going to merge what we have, but will keep an eye if this becomes an issue.

**Comment by VSadov** [comment]
Thanks!!

**Comment by VSadov** [comment]
@agocke - Is there something that need to be done to make this published/used by the runtime or it is automatic and just a matter of codeflow latency?

**Comment by agocke** [comment]
It should just move forwards. I'll keep an eye out for the darc update

**Comment by agocke** [comment]
Looks like flow is currently blocked on some unrelated changes: https://github.com/dotnet/runtime/pull/87522
